File: web/src/services/cipher_keys.ts
// Copyright 2023 OpenObserve Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

import http from "./http";

interface AkeylessStore {
  store: {
    akeyless: {
      access_id: string;
      base_url: string;
      auth: {
        type: string;
        access_key: string;
        ldap: {
          username: string;
          password: string;
        };
      };
      store: {
        type: string;
        static_secret: string;
        dfc: {
          name: string;
          iv: string;
          encrypted_data: string;
        };
      };
    };
  };
}

interface CipherKeyData {
  name: string;
  key: AkeylessStore;
  provider: Object;
}
const cipherKeys = {
  create: (org_identifier: string, data: CipherKeyData) => {
    return http().post(`/api/${org_identifier}/cipher_keys`, data);
  },
  update: (org_identifier: string, data: CipherKeyData, name: String) => {
    return http().put(`/api/${org_identifier}/cipher_keys/${name}`, data);
  },
  list: (org_identifier: string) => {
    return http().get(`/api/${org_identifier}/cipher_keys`);
  },
  get_by_name: (org_identifier: string, name: string) => {
    return http().get(`/api/${org_identifier}/cipher_keys/${name}`);
  },
  delete: (org_identifier: string, name: string) => {
    return http().delete(`/api/${org_identifier}/cipher_keys/${name}`);
  },
  bulkDelete: (org_identifier: string, data: any) => {
    return http().delete(`/api/${org_identifier}/cipher_keys/bulk`, { data });
  },
};

export default cipherKeys;

File: src/super_cluster_queue/cipher_keys.rs
// Copyright 2025 OpenObserve Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

use infra::{
    coordinator::get_coordinator,
    errors::{DbError, Error, Result},
    table::cipher::EntryKind,
};
use o2_enterprise::enterprise::super_cluster::queue::{KeysMessage, Message, MessageType};

pub(crate) async fn process(msg: Message) -> Result<()> {
    match msg.message_type {
        MessageType::KeysTable => {
            let actual_message: KeysMessage = serde_json::from_slice(&msg.value.unwrap())?;
            match actual_message {
                KeysMessage::Delete { name, org } => {
                    log::info!("[SUPER_CLUSTER:DB] deleting key {org}/{name}");
                    infra::table::cipher::remove(&org, EntryKind::CipherKey, &name).await?;
                    let cluster_coordinator = get_coordinator().await;
                    cluster_coordinator
                        .delete(&format!("/cipher_keys/{org}/{name}"), false, true, None)
                        .await?;
                }
                KeysMessage::Put { entry } => {
                    log::info!("[SUPER_CLUSTER:DB] adding key {}/{}", entry.org, entry.name);
                    let org = entry.org.clone();
                    let name = entry.name.clone();
                    match infra::table::cipher::add(entry).await {
                        Ok(_) => {}
                        // this is the case when the cluster sending the message also receives
                        // and processes it. Because message will be delivered to all clients,
                        // clients in the sending cluster will also get this, and we want to ignore
                        // that case.
                        Err(Error::DbError(DbError::UniqueViolation)) => {}
                        Err(e) => return Err(e),
                    };
                    let cluster_coordinator = get_coordinator().await;
                    cluster_coordinator
                        .put(
                            &format!("/cipher_keys/{org}/{name}"),
                            bytes::Bytes::new(), // no actual data, the receiver can query the db
                            true,
                            None,
                        )
                        .await?;
                }
                KeysMessage::Update { entry } => {
                    log::info!(
                        "[SUPER_CLUSTER:DB] updating key {}/{}",
                        entry.org,
                        entry.name
                    );
                    let org = entry.org.clone();
                    let name = entry.name.clone();
                    infra::table::cipher::update(entry).await?;
                    let cluster_coordinator = get_coordinator().await;
                    cluster_coordinator
                        .put(
                            &format!("/cipher_keys/{org}/{name}"),
                            bytes::Bytes::new(), // no actual data, the receiver can query the db
                            true,
                            None,
                        )
                        .await?;
                }
            }
        }
        _ => {
            log::error!(
                "[SUPER_CLUSTER:DB] Invalid message: type: {:?}, key: {}",
                msg.message_type,
                msg.key
            );
            return Err(Error::Message("Invalid message type".to_string()));
        }
    }
    Ok(())
}
